Source/discord/discord.cpp

```diff
@@ -0,0 +1,176 @@
+} // namespace
+
+DiscordManager::DiscordManager()
+{
+	discord::Result result = discord::Core::Create(DiscordDevilutionxAppId, DiscordCreateFlags_NoRequireDiscord, &core_);
+	good_ = result == discord::Result::Ok;
+}
+
+DiscordManager &DiscordManager::Instance()
+{
+	static DiscordManager instance {};
+	return instance;
+}
+
+std::string DiscordManager::GetLocationString() const
+{
+	const char *dungeonStr = _(/* TRANSLATORS: type of dungeon (i.e. Cathedral, Caves)*/ "None");
+	if (dungeon_area_ != DTYPE_NONE) {
+		dungeonStr = _(DungeonStrs.at(dungeon_area_));
+	}
+
+	if (dungeon_level_ > 0) {
+		int level = dungeon_level_;
+		if (dungeon_area_ == DTYPE_NEST)
+			level -= 16;
+		else if (dungeon_area_ == DTYPE_CRYPT)
+			level -= 20;
+
+		return fmt::format(_(/* TRANSLATORS: dungeon type and floor number i.e. "Cathedral 3"*/ "{} {}"), dungeonStr, level);
+	}
+	return fmt::format("{}", dungeonStr);
+}
+
+std::string DiscordManager::GetCharacterString() const
+{
+	const char *charClassStr = ClassStrTbl.at(static_cast<int>(player_class_));
+	return fmt::format(_(/* TRANSLATORS: Discord character, i.e. "Lv 6 Warrior" */ "Lv {} {}"), player_level_, charClassStr);
+}
+
+std::string DiscordManager::GetDetailString() const
+{
+	return fmt::format("{} - {}", GetCharacterString(), GetLocationString());
+}
+
+std::string DiscordManager::GetStateString() const
+{
+	return fmt::format(_(/* TRANSLATORS: Discord state i.e. "Nightmare difficulty" */ "{} difficulty"), _(DifficultyStrs.at(difficulty_)));
+}
+
+std::string DiscordManager::GetTooltipString() const
+{
+	return fmt::format("{} - {}", character_name_, GetCharacterString());
+}
+
+char DiscordManager::HeroClassChar() const
+{
+	char chr = CharChar.at(static_cast<int>(player_class_));
+	return static_cast<char>(std::tolower(chr));
+}
+
+char DiscordManager::ArmourClassChar() const
+{
+	char chr = ArmourChar.at(player_gfx_ >> 4);
+	return static_cast<char>(std::tolower(chr));
+}
+
+char DiscordManager::WpnConfigChar() const
+{
+	char chr = WepChar.at(player_gfx_ & 0xF);
+	return static_cast<char>(std::tolower(chr));
+}
+
+std::string DiscordManager::GetPlayerAssetString() const
+{
+	return fmt::format("{}{}{}as", HeroClassChar(), ArmourClassChar(), WpnConfigChar());
+}
+
+void DiscordManager::UpdateGame()
+{
+	if (core_ == nullptr || !good_)
+		return;
+
+	auto newData = std::make_tuple(Players[MyPlayerId]._pLevel, leveltype, currlevel, Players[MyPlayerId]._pgfxnum);
+	auto trackedData = std::tie(player_level_, dungeon_area_, dungeon_level_, player_gfx_);
+	if (newData != trackedData) {
+		trackedData = newData;
+
+		// Update status strings
+		discord::Activity activity = {};
+		activity.SetName(gszProductName);
+		activity.SetState(GetStateString().c_str());
+		activity.SetDetails(GetDetailString().c_str());
+		activity.SetInstance(true);
+
+		activity.GetTimestamps().SetStart(start_time_);
+
+		// Set image assets
+		activity.GetAssets().SetLargeImage(GetPlayerAssetString().c_str());

```

Comment:
`Players[MyPlayerId].AnimationData[static_cast<std::size_t>(player_graphic::Stand)]` should be initialised with the stand sprite representing the current player character class and loadout, does Discord's API allow using the sprite data directly instead of reloading it?

Reply:
It looks like `AnimationData` is for actual image data rather than the lookup string.

Discord can only use pre-uploaded sprites using string keys as lookup. I'm using the same strings that Diablo uses for this key. There are 108 character graphics (9 wpns * 3 armours * 4 characters), 1 icon, and 10 currently unused transition backgrounds, out of 300 available asset slots in the developer portal.

Reply:
That makes sense. Also explains why the string is lowercase when the mpq filenames are uppercase.